feat: add C implementation for `stats/base/dists/planck/logpmf`

yuvimittal opened the pull request:
Resolves #5000 

## Description

> What is the purpose of this pull request?

This pull request:

Add C implementation for @stdlib/stats/base/dists/planck/logpmf

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #5000 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

yuvimittal committed:
feat: Add C implementation for @stdlib/stats/base/dists/planck/logpmf

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/README.md:
```diff
@@ -101,7 +101,7 @@ y = mylogpmf( 1.0 );
 ## Notes
 
--   In virtually all cases, using the `logpmf` or `logcdf` functions is preferable to manually computing the logarithm of the `pmf` or `cdf`, respectively, since the latter is prone to overflow and underflow.
+-   In virtually all cases, using the `logpmf` or `loglogpmf` functions is preferable to manually computing the logarithm of the `pmf` or `logpmf`, respectively, since the latter is prone to overflow and underflow.
```
I think this code block is not meant to be tampered with for the following PR

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/README.md:
```diff
@@ -133,6 +133,104 @@ for ( i = 0; i < lambda.length; i++ ) {
+
+#### stdlib_base_dists_planck_logpmf( x, lambda )
+
+Evaluates the logarithm of the [probability mass function][pmf] (PMF) of a Planck (discrete exponential) distribution with shape parameter `lambda`.
```
```suggestion
Evaluates the logarithm of the probability mass function (PMF) of a Planck distribution with shape parameter `lambda`.
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/manifest.json:
```diff
@@ -0,0 +1,92 @@
+          "@stdlib/math/base/assert/is-nan",
+          "@stdlib/math/base/assert/is-nonnegative-integer",
+          "@stdlib/math/base/special/expm1",
+          "@stdlib/constants/float64/pi",
```
`pi` should be removed as it is not used in the `main.c` file right.

Neerajpathak07 replied:
same goes for the other builds as well!

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/benchmark/benchmark.js:
```diff
@@ -21,8 +21,10 @@
 
 var bench = require( '@stdlib/bench' );
+var Float64Array = require( '@stdlib/array/float64' );
 var discreteUniform = require( '@stdlib/random/array/discrete-uniform' );
```
```suggestion

```
should be removed as it is not used in the file anywhere

Neerajpathak07 replied:
same comment for the native file as well

yuvimittal replied:
actually it is been used in pkg+':factory'(benchmark 2) , as it  uses random/array/discrete-uniform , so i made two variable one for array and one for base .

Neerajpathak07 replied:
Yeah, but the the `::factory` code block should have the `discreteUniformbase` number generation logic for `len = 100` and `var x`  having a `Float64array` val just as you did it for the above code block!!

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/include/stdlib/stats/base/dists/planck/logpmf.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+/**
+* Evaluates the logpmf for an planck distribution.
```
```suggestion
* Evaluates the logarithm of the probability mass function (PMF) of a Planck distribution with shape parameter `lambda`.
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/src/main.c:
```diff
@@ -0,0 +1,49 @@
+#include "stdlib/constants/float64/ninf.h"
+
+/**
+* Evaluates the logarithm of the probability mass function (PMF) for a Planck distribution with shape parameter `lambda` at a value `x`.
```
```suggestion
* Evaluates the logarithm of the probability mass function (PMF) of a Planck distribution with shape parameter `lambda`.
```

Neerajpathak07 commented:
@yuvi-mittal 
Please change the PR title to:-
feat: add C implementation for `stats/base/dists/planck/logpmf`

yuvimittal committed:
documentation update

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

stdlib-bot committed:
chore: update copyright years

yuvimittal commented:
@Planeshifter , this PR is ready to be merged too i believe !

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/lib/native.js:
```diff
@@ -0,0 +1,67 @@
+* @private
+* @param {number} x - input value
+* @param {number} lambda - shape parameter
+* @returns {number} evaluated log(PMF)
```
```suggestion
* @returns {number} evaluated logPMF
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/src/main.c:
```diff
@@ -0,0 +1,49 @@
+*
+* @param x       input value
+* @param lambda  shape parameter
+* @return        evaluated log PMF
```
```suggestion
* @return        evaluated logPMF
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/README.md:
```diff
@@ -101,7 +101,7 @@ y = mylogpmf( 1.0 );
 ## Notes
 
--   In virtually all cases, using the `logpmf` or `logcdf` functions is preferable to manually computing the logarithm of the `pmf` or `cdf`, respectively, since the latter is prone to overflow and underflow.
+In virtually all cases, using the `logpmf` or `logcdf` functions is preferable to manually computing the logarithm of the `pmf` or `cdf`, respectively, since the latter is prone to overflow and underflow.
```
```suggestion
-   In virtually all cases, using the `logpmf` or `logcdf` functions is preferable to manually computing the logarithm of the `pmf` or `cdf`, respectively, since the latter is prone to overflow and underflow.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/README.md:
````diff
@@ -133,6 +133,104 @@ for ( i = 0; i < lambda.length; i++ ) {
+```c
+#include "stdlib/stats/base/dists/planck/logpmf.h"
+#include <stdlib.h>
+#include <stdio.h>
+static int random_discrete_uniform( const int min, const int max ) {
+    return min + ( rand() % ( max - min + 1 ) );
+}
````
```suggestion
#include <stdio.h>

static int random_discrete_uniform( const int min, const int max ) {
    return min + ( rand() % ( max - min + 1 ) );
}
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,71 @@
+	x = new Float64Array( len );
+	lambda = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		x[i] = discreteUniformBase(0, 40);
+		lambda[i] = uniformBase(1.0, 10.0);
```
```suggestion
		x[i] = discreteUniformBase( 0, 40 );
		lambda[i] = uniformBase( 1.0, 10.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,146 @@
+* @param total     total number of tests
+* @param passing   total number of passing tests
+*/
+
```
Shouldn't have empty lines between documentation blocks and functions in this file.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/logpmf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,146 @@
+	return (double)now.tv_sec + (double)now.tv_usec/1.0e6;
+}
+
+/**
+* Generates a random number on the interval [min,max).
+*
+* @param min    minimum value (inclusive)
+* @param max    maximum value (exclusive)
+* @return       random number
+*/
+
+static int random_discrete_uniform( int min, int max ) {
+	return min + rand() % ( max - min + 1 );
+}
+
+static double random_uniform( double min, double max ) {
+	double v = (double)rand() / ( (double)RAND_MAX + 1.0 );
+	return min + ( v * ( max - min ) );
+}
```
Need to document both functions here.

Planeshifter reviewed (changes requested):
Thanks for your efforts, @yuvi-mittal! This should be fairly close, but there are a bunch of errors that need to be resolved before it can be merged.

yuvimittal commented:
I will check them by EOD, but I have one question—why aren't these linting errors showing on my local system? There are no lint errors locally.

Planeshifter commented:
@yuvi-mittal Thanks for taking care of these! All checks are green on CI too, so there is nothing wrong with your local setup. Simple explanation is that we have various conventions that are currently not yet enforced via automated linting because we didn't have the bandwidth to do so or which would be infeasible to automatically check.

yuvimittal commented:
@kgryte , The PR looks good to me , can you please review if there is any issue !

Planeshifter commented:
@yuvi-mittal The comments I left still need to be addressed.

yuvimittal commented:
Yess , Sorry . I will look into that !

kgryte commented:
Ref: https://github.com/stdlib-js/stdlib/commit/32535910933dcb61a8b0cb47ac76cc789aeff6e0
